### DISCUSSION TITLE
v4.0.0 - pre-release

### DISCUSSION THREAD

**Discussion by jsjoeio** [coder/code-server#4603] [Announcements]
v4.0.0 (pre-release)

VS Code v1.63.0

Upgrading is as easy as installing the new version over the old one. code-server
maintains all user data in `~/.local/share/code-server` so that it is preserved in between
installations.

code-server has been rebased on upstream's newly open-sourced server
implementation (#4414).

### Changed

- Web socket compression has been made the default (when supported). This means
  the `--enable` flag will no longer take `permessage-deflate` as an option.
- Extra extension directories have been removed. The `--extra-extensions-dir`
  and `--extra-builtin-extensions-dir` will no longer be accepted.
- The `--install-source` and `--locale` flags have been removed.
- The static endpoint can no longer reach outside code-server. However the
  vscode-remote-resource endpoint still can.
- OpenVSX has been made the default marketplace. However this means web
  extensions like Vim may be broken.

### Deprecated

- `--link` is now deprecated (#4562).

### Breaking Changes

- You can no longer change the display language. This has been removed upstream but we hope to see if we can patch it again. Tracking in #4598

Cheers! 🍻

<hr /><em>This discussion was created from the release <a href='https://github.com/cdr/code-server/releases/tag/v4.0.0'>v4.0.0 - pre-release</a>.</em>

**Comment by jsjoeio** [comment]
Big thank you to @benz0li @drmrbrewer and others for testing this out and reporting issues! We're tracking in [this new milestone](https://github.com/cdr/code-server/milestone/34).

The goal is for myself and @code-asher to fix as many as possible before Friday. We'd like to promote 4.0.0 from a pre-release to stable on Friday because the weeks after we are out for the holidays.

We'll share more details here though. 

In the meantime, if you'd like to contribute to getting 4.0.0, here's where we could use your help:
- look at the issues in [4.0.0 milestone](https://github.com/cdr/code-server/milestone/34). See if the repro steps are as clear as possible. See if you can reproduce the issues yourself or see if you can figure out what might be causing them
- test other features on 4.0.0 and report any issues you find

Thanks all for your contributions and patience! We really appreciate it.

**Comment by jsjoeio** [comment]
## Postponing code-server 4.0.0 release + Next Version :arrow_right:
Originally we were aiming to publish the next release of code-server today. We underestimated how much we'd be able to accomplish by today. We (myself and @code-asher) agreed that it wouldn't be beneficial to everyone in the community for us to push this out. 

The majority of issues are fixed and code-server v4.0.0 prerelease is usable, but one major issue exists: you can't configure the Display Language. This means non-English speaking users wouldn't be able to change the language. Given the global community we have, we feel this (plus the other couple issues) should be fixed before we release.

### Next version - 4.0.0 -> 4.0.1
We realized we didn't use the right version when doing the pre-release. We should have called it `4.0.0-rc-1` or something similar to indicate it was a pre-release. Instead we called it `4.0.0`. This means if we released another version also called `4.0.0`, we wouldn't be able to tell which version someone was using since they'd both say `4.0.0` in the `package.json`. 

Therefore, we've decided the next release (which we'll push to all platforms like npm/Docker/etc) will be `4.0.1`. If you have questions, comments, concerns, please reply to this comment.

### F.A.Q.

#### Where can I follow along to see progress?

Take a look at the 4.0.1 milestone [here](https://github.com/coder/code-server/milestone/34). These are the issues we need to fix before we can release. 

#### Why is the release date January 7th, 2022?

We'll be on vacation and enjoying the holidays for [most of the coming weeks](https://github.com/coder/code-server/issues/4649) and don't think we can realistically get something out before then. We'd rather error on the side of caution.

#### How can I help?

Take a look at the issues in the 4.0.1 milestone. If you have time to help debug or take a look, we welcome your help! The [latest CONTRIBUTING docs](https://coder.com/docs/code-server/latest/CONTRIBUTING#development-workflow) explain how to get code-server running locally.

Thank you for your patience and understanding!

**Comment by lwmacct** [comment]
Great. I'm looking forward to it
